### PR TITLE
[FIX] web, *: keep distinct cached translation bundles per front-end lang

### PR DESCRIPTION
**Pull request by bso-odoo**
*: website_sale_stock

When translations are loaded in a front-end page, the language is
initially deduced by the server. The `GET` that obtains them does not
have a language parameter set. Because of this, this initial result is
reused from the browser cache on subsequent page accesses.
This is a problem when the user toggles between languages because the
cached content is reused while it is related to a previously selected
language.

This commit avoids this problem by making the cache specific to the
current front-end language.

Steps to reproduce:
- Make sure "Disable cache" is not enabled in your browser.
- Install website_sale_stock.
- Add the French language and activate it on the first website.
- In the backend, go to a product (e.g. Office Lamp)'s Sales tab.
- Disable "Out-of-Stock: Continue selling".
- Enter an "Out-of-Stock Message" in both English and French.
- Open the product's website page in French.
- Copy the URL.
- Open a new browser private navigation window.
- Go to the copied URL.
- Switch to English.
- Switch to French.

=> The price is first displayed correctly in French with a "," decimal
separator then is replaced by using the English "." decimal separator
and the out-of-stock message is displayed in English.

opw-3293403
opw-3343614

### CODE REVIEW

**Comment by bso-odoo** [comment]
@rdeodoo Added test tour in `website_sale_stock`.

File: addons/website_sale_stock/tests/test_website_sale_stock_multilang.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+from odoo.tests.common import HttpCase
+
+
+@tagged('post_install', '-at_install')
+class TestWebsiteSaleStockMultilang(HttpCase):
+    def test_website_sale_stock_multilang(self):
+        # Install French
+        website = self.env.ref('website.default_website')
+        lang_fr = self.env['res.lang']._activate_lang('fr_FR')
+        # lang_fr.write({'url_code': 'fr'})
```
Comment (Demesmaeker) [inline addons/website_sale_stock/tests/test_website_sale_stock_multilang.py:14]:
```suggestion
```

File: addons/website_sale_stock/tests/test_website_sale_stock_multilang.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        website.language_ids = self.env.ref('base.lang_en') + lang_fr
+        website.default_lang_id = self.env.ref('base.lang_en')
+        self.env['ir.module.module']._load_module_terms(
+            ['website', 'website_sale', 'website_sale_stock'],
+            ['en_US', 'fr_FR'], overwrite=True)
```
Comment (Demesmaeker) [inline addons/website_sale_stock/tests/test_website_sale_stock_multilang.py:19]:
```suggestion
        self.env['ir.module.module']._load_module_terms(
            ['website', 'website_sale', 'website_sale_stock'], ['en_US', 'fr_FR'], overwrite=True
        )
```

File: addons/website_sale_stock/static/tests/tours/website_sale_stock_multilang.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+}, {
+    content: "Open unavailable product page",
+    trigger: 'a[content="unavailable_product"]',
+}, {
+    content: "Check out of stock message",
+    trigger: '#out_of_stock_message:contains("Hors-stock")',
+    run: () => {}, // This is a check.
+}, {
+    content: "Check notification message",
+    trigger: '#product_stock_notification_message:contains("Soyez averti")',
```
Comment (Demesmaeker) [inline addons/website_sale_stock/static/tests/tours/website_sale_stock_multilang.js:22]: This French sentence isn't in the test data. What if tomorrow someone correct the po file with "Soyez prévenu" or anything else? I don't think the runbot will run before automatically updating a translation... :thinking: Might need some confirmation

**Review by Demesmaeker — COMMENTED** [review]
Hello!

Please keep in mind that in 16.3, in the tours, you can now write `isCheck: true` in every step that only make a check instead of `run: () => {}`

Have a nice day!

File: addons/website_sale_stock/static/tests/tours/website_sale_stock_multilang.js (hunk 3 above)
Comment (Demesmaeker) [reply addons/website_sale_stock/static/tests/tours/website_sale_stock_multilang.js]: Fine by me. This lgtm, but I was told this PR is for logistic, so I'll let them manage this =] Have a nice day!

**Comment by bso-odoo** [comment]
@rdeodoo Ready for review

File: addons/web/static/src/legacy/js/core/translation.js [hunk 4]
```diff
@@ -73,6 +73,9 @@ var TranslationDataBase = Class.extend(/** @lends instance.TranslationDataBase#
         }
         if (lang) {
             paramsGet.lang = lang;
+        } else if (session.is_frontend && session.lang_url_code) {
+            // Keep distinct cached responses per language.
+            paramsGet.unique = session.lang_url_code;
```
Comment (rdeodoo) [inline addons/web/static/src/legacy/js/core/translation.js:78]: @odoo/rd-framework-js Do you want to review this or should we proceed internally?
(The rest of the PR is just a test tour, this is the real codebase change of the PR)

Comment (ged-odoo) [reply addons/web/static/src/legacy/js/core/translation.js:78]: @rdeodoo it looks acceptable to me. I think you can proceed internally

File: addons/website_sale_stock/tests/test_website_sale_stock_multilang.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+@tagged('post_install', '-at_install')
+class TestWebsiteSaleStockMultilang(HttpCase):
+    def test_website_sale_stock_multilang(self):
+        # Install French
+        website = self.env.ref('website.default_website')
+        lang_fr = self.env['res.lang']._activate_lang('fr_FR')
+        website.language_ids = self.env.ref('base.lang_en') + lang_fr
+        website.default_lang_id = self.env.ref('base.lang_en')
+        self.env['ir.module.module']._load_module_terms(
+            ['website', 'website_sale', 'website_sale_stock'], ['en_US', 'fr_FR'], overwrite=True
```
Comment (rdeodoo) [inline addons/website_sale_stock/tests/test_website_sale_stock_multilang.py:17]:
```suggestion
            ['website', 'website_sale', 'website_sale_stock'], ['en_US', 'fr_FR'], overwrite=True,
```
And probably
```suggestion
            ['website', 'website_sale', 'website_sale_stock'],
            ['en_US', 'fr_FR'], overwrite=True,
```

File: addons/website_sale_stock/tests/test_website_sale_stock_multilang.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+
+@tagged('post_install', '-at_install')
+class TestWebsiteSaleStockMultilang(HttpCase):
+    def test_website_sale_stock_multilang(self):
+        # Install French
+        website = self.env.ref('website.default_website')
+        lang_fr = self.env['res.lang']._activate_lang('fr_FR')
+        website.language_ids = self.env.ref('base.lang_en') + lang_fr
+        website.default_lang_id = self.env.ref('base.lang_en')
```
Comment (rdeodoo) [inline addons/website_sale_stock/tests/test_website_sale_stock_multilang.py:15]: Just this should be enough?
```suggestion
        website.language_ids = [Command.link(lang_fr.id)]
```

File: addons/website_sale_stock/tests/test_website_sale_stock_multilang.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        self.env['ir.module.module']._load_module_terms(
```
Comment (rdeodoo) [inline addons/website_sale_stock/tests/test_website_sale_stock_multilang.py:16]: Why do we need this?
We can't just create the translation for the product out of stock message?

File: addons/website_sale_stock/static/tests/tours/website_sale_stock_multilang.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import tour from 'web_tour.tour';
+
+tour.register('website_sale_stock_multilang', {
+    test: true,
+    url: '/fr/shop?search=unavailable',
+},
+[{
+    content: "Do nothing",
+    trigger: "a",
+    run: () => {}, // This is a check.
```
Comment (rdeodoo) [inline addons/website_sale_stock/static/tests/tours/website_sale_stock_multilang.js:12]: ?
That's probably not needed, and if it is, we really need to know why -> comment

**Review by rdeodoo — CHANGES_REQUESTED** [review]
I was going to r+ but I actually a few questions then we can :+1: 

Also, I know I am the one who told you to write the test the easiest way and it was fine in stock, but it was a mistake, given the nature / critical aspect of the bug and the fix.
I am not confortable with having it in another module, we should have written it in the `test_website` if `website` didn't have the models/views to replicate the issue.

But let's not wait too much time now that the test is written, leave it there :+1: 

Now also fixing this new opw, could you add it in the commit message: https://www.odoo.com/web#id=3343614&cids=1&model=project.task&view_type=form
I tested the patch on their database in chrome debugger and it fixed the issue too :+1:

File: addons/website_sale_stock/tests/test_website_sale_stock_multilang.py (hunk 7 above)
Comment (bso-odoo) [reply addons/website_sale_stock/tests/test_website_sale_stock_multilang.py]: Indeed, now that the translation is hard-coded in the test, there is not need to load the module terms anymore.

**Comment by bso-odoo** [comment]
@rdeodoo Recommended changes have been applied.

File: addons/website_sale_stock/static/tests/tours/website_sale_stock_multilang.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    trigger: 'a[content="unavailable_product"]',
+}, {
+    content: "Check out of stock message",
+    trigger: '#out_of_stock_message:contains("Hors-stock")',
+    run: () => {}, // This is a check.
+}, {
+    content: "Check price",
+    trigger: 'span:contains("123,45")',
+    run: () => {}, // This is a check.
+}, {
```
Comment (rdeodoo) [inline addons/website_sale_stock/static/tests/tours/website_sale_stock_multilang.js:20]: Note that `extra_trigger` is generally a good fit to merge 2 check steps (same twice later)

**Review by rdeodoo — APPROVED** [review]
@robodoo r+
Thank you :+1:

**Pull request closed without merging**